File: KeeperLegacyTests/EconomyTests.swift
import XCTest
@testable import KeeperLegacy

// MARK: - Economy Tests
// Tests for pricing, sell values, and coin transactions.
// These run on every GitHub Actions build to catch economy balance regressions.

final class EconomyTests: XCTestCase {

    // MARK: Pricing Table

    func testCommonCreatureBasePrice() {
        XCTAssertEqual(Rarity.common.basePrice, 150)
    }

    func testRarityPriceOrdering() {
        XCTAssertLessThan(Rarity.common.basePrice, Rarity.uncommon.basePrice)
        XCTAssertLessThan(Rarity.uncommon.basePrice, Rarity.rare.basePrice)
    }

    func testRaritySellValueOrdering() {
        XCTAssertLessThan(Rarity.common.baseSellValue, Rarity.uncommon.baseSellValue)
        XCTAssertLessThan(Rarity.uncommon.baseSellValue, Rarity.rare.baseSellValue)
    }

    func testSellValueWithNeutralHappiness() {
        // At 0.0 happiness multiplier (unhappy) → minimum 1.0x
        let value = PricingTable.sellValue(rarity: .common, happinessMultiplier: 1.0)
        XCTAssertEqual(value, Rarity.common.baseSellValue)
    }

    func testSellValueWithMaxHappiness() {
        // At 2.0 multiplier (max happy) → 2x base
        let value = PricingTable.sellValue(rarity: .common, happinessMultiplier: 2.0)
        XCTAssertEqual(value, Rarity.common.baseSellValue * 2)
    }

    func testSellValueClampsAboveMax() {
        // Multiplier > 2.0 should be clamped to 2.0
        let capped   = PricingTable.sellValue(rarity: .common, happinessMultiplier: 5.0)
        let atMax    = PricingTable.sellValue(rarity: .common, happinessMultiplier: 2.0)
        XCTAssertEqual(capped, atMax)
    }

    // MARK: Player Economy

    func testEarnCoins() {
        var economy = PlayerEconomy(startingCoins: 100)
        economy.earnCoins(50)
        XCTAssertEqual(economy.coins, 150)
        XCTAssertEqual(economy.totalCoinsEarned, 150)
    }

    func testSpendCoinsSuccess() {
        var economy = PlayerEconomy(startingCoins: 200)
        let result = economy.spendCoins(100)
        XCTAssertTrue(result)
        XCTAssertEqual(economy.coins, 100)
        XCTAssertEqual(economy.totalCoinsSpent, 100)
    }

    func testSpendCoinsInsufficientFunds() {
        var economy = PlayerEconomy(startingCoins: 50)
        let result = economy.spendCoins(100)
        XCTAssertFalse(result)
        XCTAssertEqual(economy.coins, 50)  // Unchanged
    }

    func testCannotGoNegative() {
        var economy = PlayerEconomy(startingCoins: 10)
        economy.spendCoins(10)
        let result = economy.spendCoins(1)
        XCTAssertFalse(result)
        XCTAssertEqual(economy.coins, 0)
    }

    // MARK: Breeding Costs

    func testBreedingCostOrdering() {
        let common   = PricingTable.Breeding.cost(rarity: .common)
        let uncommon = PricingTable.Breeding.cost(rarity: .uncommon)
        let rare     = PricingTable.Breeding.cost(rarity: .rare)
        XCTAssertLessThan(common, uncommon)
        XCTAssertLessThan(uncommon, rare)
    }

    // MARK: Food Catalog

    func testFoodCatalogNotEmpty() {
        XCTAssertFalse(PricingTable.Food.catalog.isEmpty)
    }

    func testFoodHungerRestoreRange() {
        for food in PricingTable.Food.catalog {
            XCTAssertGreaterThan(food.hungerRestored, 0.0, "\(food.name) restores nothing")
            XCTAssertLessThanOrEqual(food.hungerRestored, 1.0, "\(food.name) restores over 100%")
        }
    }

    func testFoodCostPositive() {
        for food in PricingTable.Food.catalog {
            XCTAssertGreaterThan(food.coinCost, 0, "\(food.name) has zero cost")
        }
    }
}

File: KeeperLegacyGodot/Tests/EconomyTests.cs
// Tests/EconomyTests.cs
// Port of EconomyTests.swift

using NUnit.Framework;
using KeeperLegacy.Models;

namespace KeeperLegacy.Tests
{
    [TestFixture]
    public class EconomyTests
    {
        // ── Pricing Table ─────────────────────────────────────────────────────

        [Test] public void CommonCreatureBasePrice()
            => Assert.That(Rarity.Common.BasePrice(), Is.EqualTo(150));

        [Test]
        public void RarityPriceOrdering()
        {
            Assert.That(Rarity.Common.BasePrice(),   Is.LessThan(Rarity.Uncommon.BasePrice()));
            Assert.That(Rarity.Uncommon.BasePrice(), Is.LessThan(Rarity.Rare.BasePrice()));
        }

        [Test]
        public void RaritySellValueOrdering()
        {
            Assert.That(Rarity.Common.BaseSellValue(),   Is.LessThan(Rarity.Uncommon.BaseSellValue()));
            Assert.That(Rarity.Uncommon.BaseSellValue(), Is.LessThan(Rarity.Rare.BaseSellValue()));
        }

        [Test]
        public void SellValueWithNeutralHappiness()
        {
            int value = PricingTable.SellValue(Rarity.Common, happinessMultiplier: 1.0);
            Assert.That(value, Is.EqualTo(Rarity.Common.BaseSellValue()));
        }

        [Test]
        public void SellValueWithMaxHappiness()
        {
            int value = PricingTable.SellValue(Rarity.Common, happinessMultiplier: 2.0);
            Assert.That(value, Is.EqualTo(Rarity.Common.BaseSellValue() * 2));
        }

        [Test]
        public void SellValueClampsAboveMax()
        {
            int capped = PricingTable.SellValue(Rarity.Common, happinessMultiplier: 5.0);
            int atMax  = PricingTable.SellValue(Rarity.Common, happinessMultiplier: 2.0);
            Assert.That(capped, Is.EqualTo(atMax));
        }

        // ── Player Economy ────────────────────────────────────────────────────

        [Test]
        public void EarnCoins()
        {
            var e = new PlayerEconomy(startingCoins: 100);
            e.EarnCoins(50);
            Assert.That(e.Coins,           Is.EqualTo(150));
            Assert.That(e.TotalCoinsEarned, Is.EqualTo(150));
        }

        [Test]
        public void SpendCoinsSuccess()
        {
            var e = new PlayerEconomy(startingCoins: 200);
            bool result = e.SpendCoins(100);
            Assert.That(result,           Is.True);
            Assert.That(e.Coins,          Is.EqualTo(100));
            Assert.That(e.TotalCoinsSpent, Is.EqualTo(100));
        }

        [Test]
        public void SpendCoinsInsufficientFunds()
        {
            var e = new PlayerEconomy(startingCoins: 50);
            bool result = e.SpendCoins(100);
            Assert.That(result,  Is.False);
            Assert.That(e.Coins, Is.EqualTo(50));
        }

        [Test]
        public void CannotGoNegative()
        {
            var e = new PlayerEconomy(startingCoins: 10);
            e.SpendCoins(10);
            bool result = e.SpendCoins(1);
            Assert.That(result,  Is.False);
            Assert.That(e.Coins, Is.EqualTo(0));
        }

        // ── Breeding Costs ────────────────────────────────────────────────────

        [Test]
        public void BreedingCostOrdering()
        {
            int common   = PricingTable.Breeding.Cost(Rarity.Common);
            int uncommon = PricingTable.Breeding.Cost(Rarity.Uncommon);
            int rare     = PricingTable.Breeding.Cost(Rarity.Rare);
            Assert.That(common,   Is.LessThan(uncommon));
            Assert.That(uncommon, Is.LessThan(rare));
        }

        // ── Food Catalog ──────────────────────────────────────────────────────

        [Test] public void FoodCatalogNotEmpty()
            => Assert.That(PricingTable.Food.Catalog, Is.Not.Empty);

        [Test]
        public void FoodHungerRestoreRange()
        {
            foreach (var food in PricingTable.Food.Catalog)
            {
                Assert.That(food.HungerRestored, Is.GreaterThan(0.0),    $"{food.Name} restores nothing");
                Assert.That(food.HungerRestored, Is.LessThanOrEqualTo(1.0), $"{food.Name} restores over 100%");
            }
        }

        [Test]
        public void FoodCostPositive()
        {
            foreach (var food in PricingTable.Food.Catalog)
                Assert.That(food.CoinCost, Is.GreaterThan(0), $"{food.Name} has zero cost");
        }
    }
}
